karabo_data/stacking.py

```diff
@@ -0,0 +1,208 @@
+        to the geometry machinery to assemble images.
+
+    Returns
+    -------
+    combined: numpy.array
+        Stacked data for requested data path.
+    """
+
+    if not train:
+        raise ValueError("No data")
+
+    dtypes, shapes, empty_mods = set(), set(), set()
+    modno_arrays = {}
+    for device in train:
+        det_mod_match = re.search(r'/DET/(\d+)CH', device)
+        if not det_mod_match:
+            raise ValueError("Non-detector source: {}".format(device))
+        modno = int(det_mod_match.group(1))
+
+        try:
+            array = train[device][data]
+        except KeyError:
+            continue
+        dtypes.add(array.dtype)
+        shapes.add(array.shape)
+        modno_arrays[modno] = array
+
+    if len(dtypes) > 1:
+        raise ValueError("Arrays have mismatched dtypes: {}".format(dtypes))
+    if len(shapes) > 1:
+        s1, s2, *_ = sorted(shapes)
+        if len(shapes) > 2 or (s1[0] != 0) or (s1[1:] != s2[1:]):
+            raise ValueError("Arrays have mismatched shapes: {}".format(shapes))
+        empty_mods = {n for n, a in modno_arrays.items() if a.shape == s1}
+        for modno in empty_mods:
+            del modno_arrays[modno]
+        shapes.remove(s1)
+    if max(modno_arrays) >= modules:
+        raise IndexError("Module {} is out of range for a detector with {} modules"
+                         .format(max(modno_arrays), modules))
+
+    dtype = dtypes.pop()
+    shape = shapes.pop()
+    stack = VirtualStack(
+        modno_arrays, modules, shape, dtype, fillvalue, stack_axis=axis
+    )
+    if virtual:
+        return stack
+
+    return stack.asarray()
+
+
+class VirtualStack:
+    """'Virtual' array for stacking detector data from several modules.
+
+    Access is limited to either a single module at a time or all modules
+    together, but this is enough to assemble detector images.
+    """
+    def __init__(self, data, nmodules, mod_shape, dtype, fillvalue,
+                 stack_axis=-3):
+        self._nmodules = nmodules
+        self._data = data  # {modno: array}
+        self.dtype = dtype
+        self._fillvalue = fillvalue
+        self._mod_shape = mod_shape
+        self.ndim = len(mod_shape) + 1
+        self._stack_axis = stack_axis
+        if self._stack_axis < 0:
+            self._stack_axis += self.ndim
+        sax = self._stack_axis
+        self.shape = mod_shape[:sax] + (nmodules,) + mod_shape[sax:]
+
+    def __repr__(self):
+        return "<VirtualStack (shape={}, {}/{} modules, dtype={})>".format(
+            self.shape, len(self._data), self._nmodules, self.dtype,
+        )
+
+    # Multidimensional slicing
+    def __getitem__(self, item):
+        if not isinstance(item, tuple):
+            item = (item,)
+
+        missing_dims = self.ndim - len(item)
+        if Ellipsis in item:
+            ix = item.index(Ellipsis)
+            missing_dims += 1
+            item = item[:ix] + (slice(None, None),) * missing_dims + item[ix+1:]
```

Comment:
Suppose you have a 4D array x and slice it with x[(0, ..., 0)]. Then `ix` will be 1, but  I guess `item` should be

```
item = item[:ix] + (slice(None, None),) * 2 + item[ix+2:]
```
You are doing something fancy and I am trying to follow you :-)

Reply:
Slicing in Python includes the first index specified but excludes the last one (I believe this is called "half-open intervals"). So `item[:1], item[1:]` would split item with no gap and no duplicate. `item[ix+1]` leaves a gap of one element from the input, which is the `...`.

Perhaps it's clearer with an example:

```
In [1]: (0, ..., 0)[:1]                                                                 
Out[1]: (0,)

In [2]: (0, ..., 0)[2:]                                                               
Out[2]: (0,)

In [3]: (0, ..., 0)[:1] + (1, 1) + (0, ..., 0)[2:]                                      
Out[3]: (0, 1, 1, 0)
```

Does that make sense?

Reply:
Thanks! I misled myself. `item` here is the slicing instead of the array itself.

Reply:
Yup. That was the default naming Pycharm gave when I created the `__getitem__` method, but maybe it would be clearer with `slices`.